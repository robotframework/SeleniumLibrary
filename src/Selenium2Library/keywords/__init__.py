from ._logging import _LoggingKeywords
from ._runonfailure import _RunOnFailureKeywords
from ._browsermanagement import _BrowserManagementKeywords
from ._element import _ElementKeywords
from ._tableelement import _TableElementKeywords
from ._formelement import _FormElementKeywords
from ._selectelement import _SelectElementKeywords
from ._javascript import _JavaScriptKeywords
from ._cookie import _CookieKeywords
from ._screenshot import _ScreenshotKeywords
from ._waiting import _WaitingKeywords
from ._alert import _AlertKeywords


__all__ = [
    "_LoggingKeywords",
    "_RunOnFailureKeywords",
    "_BrowserManagementKeywords",
    "_ElementKeywords",
    "_TableElementKeywords",
    "_FormElementKeywords",
    "_SelectElementKeywords",
    "_JavaScriptKeywords",
    "_CookieKeywords",
    "_ScreenshotKeywords",
    "_WaitingKeywords",
    "_AlertKeywords"
]
