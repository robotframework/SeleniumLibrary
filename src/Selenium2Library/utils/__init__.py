import os
import sys
from fnmatch import fnmatch
from .browsercache import BrowserCache

__all__ = [
    "get_child_packages_in",
    "get_module_names_under",
    "import_modules_under",
    "escape_xpath_value",
    "BrowserCache"
]

PY3 = sys.version_info[0] == 3
if PY3:
    unicode = str


# Public

def get_child_packages_in(root_dir, include_root_package_name=True, exclusions=None):
    packages = []
    root_package_str = os.path.basename(root_dir) + '.' if include_root_package_name else ""
    _discover_child_package_dirs(
        root_dir,
        _clean_exclusions(exclusions),
        lambda abs_path, relative_path, name: 
            packages.append(root_package_str + relative_path.replace(os.sep, '.')))
    return packages

def get_module_names_under(root_dir, include_root_package_name=True, exclusions=None, pattern=None):
    module_names = []
    root_package_str = os.path.basename(root_dir) + '.' if include_root_package_name else ""
    _discover_module_files_in(
        root_dir,
        _clean_exclusions(exclusions),
        pattern if pattern is not None else "*.*",
        lambda abs_path, relative_path, name: 
            module_names.append(root_package_str + os.path.splitext(relative_path)[0].replace(os.sep, '.')))
    return module_names

def import_modules_under(root_dir, include_root_package_name=True, exclusions=None, pattern=None):
    module_names = get_module_names_under(root_dir, include_root_package_name, exclusions, pattern)
    modules = [ __import__(module_name, globals(), locals(), ['*'], 0)
        for module_name in module_names ]
    return (module_names, modules)

def escape_xpath_value(value):
    value = unicode(value)
    if '"' in value and '\'' in value:
        parts_wo_apos = value.split('\'')
        return "concat('%s')" % "', \"'\", '".join(parts_wo_apos)
    if '\'' in value:
        return "\"%s\"" % value
    return "'%s'" % value

# Private

def _clean_exclusions(exclusions):
    if exclusions is None: exclusions = []
    if not isinstance(exclusions, list): exclusions = [ exclusions ]
    exclusions = [ os.sep + exclusion.lower().strip(os.sep) + os.sep
        for exclusion in exclusions ]
    return exclusions

def _discover_child_package_dirs(root_dir, exclusions, callback, relative_dir=None):
    relative_dir = relative_dir if relative_dir is not None else ''
    abs_dir = os.path.join(root_dir, relative_dir)
    for item in os.listdir(abs_dir):
        item_relative_path = os.path.join(relative_dir, item)
        item_abs_path = os.path.join(root_dir, item_relative_path)
        if os.path.isdir(item_abs_path):
            if os.path.exists(os.path.join(item_abs_path, "__init__.py")):
                exclusion_matches = [ exclusion for exclusion in exclusions 
                    if os.sep + item_relative_path.lower() + os.sep == exclusion ]
                if not exclusion_matches:
                    callback(item_abs_path, item_relative_path, item)
                    _discover_child_package_dirs(root_dir, exclusions, callback, item_relative_path)

def _discover_module_files_in(root_dir, exclusions, pattern, callback):
    def find_matching_files(relative_dir):
        abs_dir = os.path.join(root_dir, relative_dir)
        for item in os.listdir(abs_dir):
            item_relative_path = os.path.join(relative_dir, item)
            item_abs_path = os.path.join(root_dir, item_relative_path)
            if os.path.isfile(item_abs_path) and fnmatch(item, pattern):
                callback(item_abs_path, item_relative_path, item)

    find_matching_files('')

    _discover_child_package_dirs(
        root_dir,
        _clean_exclusions(exclusions),
        lambda abs_path, relative_path, name: find_matching_files(relative_path))
