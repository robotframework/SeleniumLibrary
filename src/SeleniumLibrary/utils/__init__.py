# Copyright 2008-2011 Nokia Networks
# Copyright 2011-2016 Ryan Tomac, Ed Manlove and contributors
# Copyright 2016-     Robot Framework Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from robot.utils import plural_or_not, secs_to_timestr, timestr_to_secs  # noqa

from .librarylistener import LibraryListener  # noqa
from .types import (
    is_falsy,
    is_noney,
    is_string,
    is_truthy,
    WINDOWS,
    _convert_timeout,
)  # noqa


def escape_xpath_value(value: str):
    value = str(value)
    if '"' in value and "'" in value:
        parts_wo_apos = value.split("'")
        escaped = "', \"'\", '".join(parts_wo_apos)
        return f"concat('{escaped}')"
    if "'" in value:
        return f'"{value}"'
    return f"'{value}'"
