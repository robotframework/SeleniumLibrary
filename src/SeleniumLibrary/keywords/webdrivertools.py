# Copyright 2008-2011 Nokia Networks
# Copyright 2011-2016 Ryan Tomac, Ed Manlove and contributors
# Copyright 2016-     Robot Framework Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import warnings

from robot.utils import ConnectionCache
from selenium import webdriver

from SeleniumLibrary.utils import is_falsy, is_truthy, SELENIUM_VERSION


class WebDriverCreator(object):

    browser_names = {
        'googlechrome': "chrome",
        'gc': "chrome",
        'chrome': "chrome",
        'headlesschrome': 'headless_chrome',
        'ff': 'firefox',
        'firefox': 'firefox',
        'headlessfirefox': 'headless_firefox',
        'ie': 'ie',
        'internetexplorer': 'ie',
        'edge': 'edge',
        'opera': 'opera',
        'safari': 'safari',
        'phantomjs': 'phantomjs',
        'htmlunit': 'htmlunit',
        'htmlunitwithjs': 'htmlunit_with_js',
        'android': 'android',
        'iphone': 'iphone'
    }

    def __init__(self, log_dir):
        self.log_dir = log_dir

    def create_driver(self, browser, desired_capabilities, remote_url,
                      profile_dir=None):
        creation_method = self._get_creator_method(browser)
        desired_capabilities = self._parse_capabilities(desired_capabilities, browser)
        if (creation_method == self.create_firefox
                or creation_method == self.create_headless_firefox):
            return creation_method(desired_capabilities, remote_url,
                                   profile_dir)
        return creation_method(desired_capabilities, remote_url)

    def _get_creator_method(self, browser):
        browser = browser.lower().replace(' ', '')
        if browser in self.browser_names:
            return getattr(self, 'create_{}'.format(self.browser_names[browser]))
        raise ValueError('{} is not a supported browser.'.format(browser))

    def _parse_capabilities(self, capabilities, browser=None):
        if is_falsy(capabilities):
            return {}
        if not isinstance(capabilities, dict):
            capabilities = self._string_to_dict(capabilities)
        browser_alias = {'googlechrome': "chrome", 'gc': "chrome",
                         'headlesschrome': 'chrome', 'ff': 'firefox',
                         'headlessfirefox': 'firefox',
                         'internetexplorer': 'ie'}
        browser = browser_alias.get(browser, browser)
        if browser in ['ie', 'firefox', 'edge']:
            return {'capabilities': capabilities}
        return {'desired_capabilities': capabilities}

    def _string_to_dict(self, capabilities):
        desired_capabilities = {}
        for part in capabilities.split(','):
            key, value = part.split(':')
            desired_capabilities[key.strip()] = value.strip()
        return desired_capabilities

    def create_chrome(self, desired_capabilities, remote_url, options=None):
        if is_truthy(remote_url):
            if not desired_capabilities:
                desired_capabilities = {'desired_capabilities': webdriver.DesiredCapabilities.CHROME.copy()}
            return self._remote(desired_capabilities, remote_url, options=options)
        if SELENIUM_VERSION.major >= 3 and SELENIUM_VERSION.minor >= 8:
            return webdriver.Chrome(options=options, **desired_capabilities)
        return webdriver.Chrome(**desired_capabilities)

    def create_headless_chrome(self, desired_capabilities, remote_url):
        if SELENIUM_VERSION.major >= 3 and SELENIUM_VERSION.minor >= 8:
            options = webdriver.ChromeOptions()
            options.set_headless()
        else:
            options = None
        return self.create_chrome(desired_capabilities, remote_url, options)

    def create_firefox(self, desired_capabilities, remote_url, ff_profile_dir,
                       options=None):
        profile = self._get_ff_profile(ff_profile_dir)
        if is_truthy(remote_url):
            if not desired_capabilities:
                desired_capabilities = {'desired_capabilities': webdriver.DesiredCapabilities.FIREFOX.copy()}
            return self._remote(desired_capabilities, remote_url,
                                profile, options)
        desired_capabilities.update(self._geckodriver_log)
        if SELENIUM_VERSION.major >= 3 and SELENIUM_VERSION.minor >= 8:
            return webdriver.Firefox(options=options, firefox_profile=profile,
                                     **desired_capabilities)
        return webdriver.Firefox(firefox_profile=profile,
                                 **desired_capabilities)

    def _get_ff_profile(self, ff_profile_dir):
        if is_falsy(ff_profile_dir):
            return webdriver.FirefoxProfile()
        return webdriver.FirefoxProfile(ff_profile_dir)

    @property
    def _geckodriver_log(self):
        if SELENIUM_VERSION.major >= 3:
            return {'log_path': os.path.join(self.log_dir, 'geckodriver.log')}
        return {}

    def create_headless_firefox(self, desired_capabilities, remote_url,
                                ff_profile_dir):
        if SELENIUM_VERSION.major >= 3 and SELENIUM_VERSION.minor >= 8:
            options = webdriver.FirefoxOptions()
            options.set_headless()
        else:
            options = None
        return self.create_firefox(desired_capabilities, remote_url,
                                   ff_profile_dir, options)

    def create_ie(self, desired_capabilities, remote_url):
        if is_truthy(remote_url):
            if not desired_capabilities:
                ie = webdriver.DesiredCapabilities.INTERNETEXPLORER.copy()
                desired_capabilities = {'desired_capabilities': ie}
            return self._remote(desired_capabilities, remote_url)
        return webdriver.Ie(**desired_capabilities)

    def create_edge(self, desired_capabilities, remote_url):
        if is_truthy(remote_url):
            if not desired_capabilities:
                edge = webdriver.DesiredCapabilities.EDGE.copy()
                desired_capabilities = {'desired_capabilities': edge}
            return self._remote(desired_capabilities, remote_url)
        return webdriver.Edge(**desired_capabilities)

    def create_opera(self, desired_capabilities, remote_url):
        if is_truthy(remote_url):
            if not desired_capabilities:
                opera = webdriver.DesiredCapabilities.OPERA.copy()
                desired_capabilities = {'desired_capabilities': opera}
            return self._remote(desired_capabilities, remote_url)
        return webdriver.Opera(**desired_capabilities)

    def create_safari(self, desired_capabilities, remote_url):
        if is_truthy(remote_url):
            if not desired_capabilities:
                caps = webdriver.DesiredCapabilities.SAFARI.copy()
                desired_capabilities = {'desired_capabilities': caps}
            return self._remote(desired_capabilities, remote_url)
        return webdriver.Safari(**desired_capabilities)

    def create_phantomjs(self, desired_capabilities, remote_url):
        warnings.warn('SeleniumLibrary support for PhantomJS has been deprecated, '
                      'please use headlesschrome or headlessfirefox instead.')
        if is_truthy(remote_url):
            if not desired_capabilities:
                caps = webdriver.DesiredCapabilities.PHANTOMJS.copy()
                desired_capabilities = {'desired_capabilities': caps}
            return self._remote(desired_capabilities, remote_url)
        return webdriver.PhantomJS(**desired_capabilities)

    def create_htmlunit(self, desired_capabilities, remote_url):
        if not desired_capabilities:
            desired_capabilities['desired_capabilities'] = webdriver.DesiredCapabilities.HTMLUNIT
        return self._remote(desired_capabilities, remote_url)

    def create_htmlunit_with_js(self, desired_capabilities, remote_url):
        if not desired_capabilities:
            desired_capabilities['desired_capabilities'] = webdriver.DesiredCapabilities.HTMLUNITWITHJS
        return self._remote(desired_capabilities, remote_url)

    def create_android(self, desired_capabilities, remote_url):
        if not desired_capabilities:
            desired_capabilities['desired_capabilities'] = webdriver.DesiredCapabilities.ANDROID
        return self._remote(desired_capabilities, remote_url)

    def create_iphone(self, desired_capabilities, remote_url):
        if not desired_capabilities:
            desired_capabilities['desired_capabilities'] = webdriver.DesiredCapabilities.IPHONE
        return self._remote(desired_capabilities, remote_url)

    def _remote(self, desired_capabilities, remote_url,
                profile_dir=None, options=None):
        remote_url = str(remote_url)
        if 'capabilities' in desired_capabilities:
            desired_capabilities['desired_capabilities'] = desired_capabilities.pop('capabilities')
        if SELENIUM_VERSION.major >= 3 and SELENIUM_VERSION.minor >= 8:
            return webdriver.Remote(command_executor=remote_url,
                                    browser_profile=profile_dir, options=options,
                                    **desired_capabilities)
        return webdriver.Remote(command_executor=remote_url,
                                browser_profile=profile_dir,
                                **desired_capabilities)


class WebDriverCache(ConnectionCache):

    def __init__(self):
        ConnectionCache.__init__(self, no_current_msg='No current browser')
        self._closed = set()

    @property
    def drivers(self):
        return self._connections

    @property
    def active_drivers(self):
        open_drivers = []
        for driver in self._connections:
            if driver not in self._closed:
                open_drivers.append(driver)
        return open_drivers

    def close(self):
        if self.current:
            driver = self.current
            driver.quit()
            self.current = self._no_current
            self._closed.add(driver)

    def close_all(self):
        for driver in self._connections:
            if driver not in self._closed:
                driver.quit()
        self.empty_cache()
        return self.current
