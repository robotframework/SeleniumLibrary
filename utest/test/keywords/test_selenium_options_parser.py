import inspect
import os
import unittest

import pytest
from approvaltests.approvals import verify_all
from approvaltests.reporters.generic_diff_reporter_factory import (
    GenericDiffReporterFactory,
)
from mockito import mock, when, unstub, ANY
from robot.utils import WINDOWS
from selenium import webdriver

from SeleniumLibrary.keywords.webdrivertools import SeleniumOptions, WebDriverCreator


@pytest.fixture(scope="module")
def options():
    return SeleniumOptions()


@pytest.fixture(scope="module")
def reporter():
    path = os.path.dirname(__file__)
    reporter_json = os.path.abspath(
        os.path.join(path, "..", "approvals_reporters.json")
    )
    factory = GenericDiffReporterFactory()
    factory.load(reporter_json)
    return factory.get_first_working()


def teardown_function():
    unstub()


@unittest.skipIf(WINDOWS, reason="ApprovalTest do not support different line feeds")
def test_parse_options_string(options, reporter):
    results = []
    results.append(options._parse('method("arg1")'))
    results.append(options._parse('method("arg1", "arg2")'))
    results.append(options._parse("method(True)"))
    results.append(options._parse("method(1)"))
    results.append(options._parse('method("arg1", 2, None, False, "arg2")'))
    results.append(options._parse('method ( " arg1 " , 2 , None , False , " arg2 " )'))
    results.append(options._parse('attribute="arg1"'))
    results.append(options._parse("  attribute = True  "))
    results.append(options._parse('method("arg1");attribute=True'))
    results.append(options._parse('method("arg1") ; attribute=True ; method("arg2")'))
    results.append(options._parse("attribute"))
    results.append(options._parse("method()"))
    results.append(options._parse("method(None)"))
    results.append(options._parse('method("--proxy 10.10.1.3:2345")'))
    results.append(options._parse('method(";arg1")'))
    results.append(options._parse('method  (   "arg1"  ,    2    ,"arg2"   )'))
    results.append(options._parse("method('arg1')"))
    results.append(
        options._parse(
            'add_argument("-profile"); add_argument("C:\\\\path\\to\\\\profile")'
        )
    )
    results.append(
        options._parse(
            r'add_argument("-profile"); add_argument("C:\\path\\to\\profile")'
        )
    )
    results.append(options._parse("attribute=None"))
    results.append(
        options._parse(
            'method("foo", {"key": False});attribute=True;method("bar", {"key": None})'
        )
    )
    verify_all("Selenium options string to dict", results, reporter=reporter)


@unittest.skipIf(WINDOWS, reason="ApprovalTest do not support different line feeds")
def test_index_of_separator(options, reporter):
    results = []
    results.append(options._get_arument_index('method({"key": "value"})'))
    results.append(options._get_arument_index('attribute={"key": "value"}'))
    results.append(options._get_arument_index('method(foo={"key": "value"})'))
    results.append(options._get_arument_index('attribute=("value1", "value2")'))
    verify_all("Get argument index", results, reporter=reporter)


@unittest.skipIf(WINDOWS, reason="ApprovalTest do not support different line feeds")
def test_parse_complex_object(options, reporter):
    results = []
    results.append(options._parse_to_tokens('method({"key": "value"})'))
    results.append(options._parse_to_tokens('attribute={"key": "value"}'))
    results.append(options._parse_to_tokens('attribute=("value1", "value2")'))
    results.append(options._parse_to_tokens('method("foo", {"key": "value"})'))
    verify_all("Parse complex Python object", results, reporter=reporter)


@unittest.skipIf(WINDOWS, reason="ApprovalTest do not support different line feeds")
def test_parse_arguemnts(options, reporter):
    results = []
    results.append(options._parse_arguments(("arg1",), True))
    results.append(options._parse_arguments("arg1", False))
    results.append(options._parse_arguments({"key": "value"}, False))
    results.append(options._parse_arguments(["value1", "value2"], False))
    results.append(options._parse_arguments(("foo", {"key": "value"}), False))
    verify_all("Parse arguments from complex object", results, reporter=reporter)


@unittest.skipIf(WINDOWS, reason="ApprovalTest do not support different line feeds")
def test_parse_options_string_errors(options, reporter):
    results = []
    results.append(error_formatter(options._parse, 'method("arg1)', True))
    results.append(error_formatter(options._parse, 'method(arg1")', True))
    results.append(error_formatter(options._parse, "method(arg1)", True))
    results.append(error_formatter(options._parse, "attribute=arg1", True))
    results.append(error_formatter(options._parse, "attribute=webdriver", True))
    results.append(error_formatter(options._parse, 'method(argument="value")', True))
    verify_all("Selenium options string errors", results, reporter=reporter)


@unittest.skipIf(WINDOWS, reason="ApprovalTest do not support different line feeds")
def test_split_options(options, reporter):
    results = []
    results.append(options._split('method("arg1");method("arg2")'))
    results.append(options._split('method("arg1")'))
    results.append(options._split("attribute=True"))
    results.append(
        options._split('attribute="semi;colons;middle";other_attribute=True')
    )
    results.append(options._split('method("arg1;");method(";arg2;")'))
    results.append(options._split(' method ( " arg1 ") ; method ( " arg2 " ) '))
    verify_all("Selenium options string splitting", results, reporter=reporter)


@unittest.skipIf(WINDOWS, reason="ApprovalTest do not support different line feeds")
def test_options_create(options, reporter):
    results = []
    options_str = 'add_argument("--disable-dev-shm-usage")'
    sel_options = options.create("chrome", options_str)
    results.append(sel_options.arguments)

    options_str = f'{options_str};add_argument("--headless")'
    sel_options = options.create("chrome", options_str)
    results.append(sel_options.arguments)

    options_str = f'{options_str};add_argument("--proxy-server=66.97.38.58:80")'
    sel_options = options.create("chrome", options_str)
    results.append(sel_options.arguments)

    options_str = f'{options_str};binary_location("too", "many", "args")'
    try:
        options.create("chrome", options_str)
    except Exception as error:
        results.append(error.__str__()[:7])

    chrome_options = webdriver.ChromeOptions()
    chrome_options.add_argument("--disable-dev-shm-usage")
    sel_options = options.create("chrome", chrome_options)
    results.append(sel_options.arguments)

    sel_options = options.create("chrome", None)
    results.append(sel_options)

    verify_all("Selenium options", results, reporter=reporter)


@unittest.skipIf(WINDOWS, reason="ApprovalTest do not support different line feeds")
def test_create_with_android(options, reporter):
    results = []
    chrome_options = webdriver.ChromeOptions()
    chrome_options.add_experimental_option("androidPackage", "com.android.chrome")
    sel_options = options.create("android", chrome_options)
    results.append([sel_options.arguments, sel_options.experimental_options])
    verify_all("Selenium options with android", results, reporter=reporter)


@unittest.skipIf(WINDOWS, reason="ApprovalTest do not support different line feeds")
def test_get_options(options, reporter):
    options_str = 'add_argument("--proxy-server=66.97.38.58:80")'
    sel_options = options.create("chrome", options_str)
    results = [sel_options.arguments]
    verify_all("Selenium options with string.", results, reporter=reporter)


@unittest.skipIf(WINDOWS, reason="ApprovalTest do not support different line feeds")
def test_importer(options, reporter):
    results = []
    results.append(options._import_options("firefox"))
    results.append(options._import_options("headless_firefox"))
    results.append(options._import_options("chrome"))
    results.append(options._import_options("headless_chrome"))
    results.append(options._import_options("ie"))
    results.append(options._import_options("opera"))
    results.append(options._import_options("edge"))
    results.append(error_formatter(options._import_options, "phantomjs"))
    results.append(error_formatter(options._import_options, "safari"))
    results.append(error_formatter(options._import_options, "htmlunit"))
    results.append(error_formatter(options._import_options, "htmlunit_with_js"))
    results.append(options._import_options("android"))
    results.append(error_formatter(options._import_options, "iphone"))
    verify_all("Selenium options import", results, reporter=reporter)


def error_formatter(method, arg, full=False):
    try:
        return method(arg)
    except Exception as error:
        if full:
            return f"{arg} {error}"
        return "{} {}".format(arg, error.__str__()[:15])


@pytest.fixture(scope="module")
def creator():
    curr_dir = os.path.dirname(os.path.abspath(__file__))
    output_dir = os.path.abspath(os.path.join(curr_dir, "..", "..", "output_dir"))
    return WebDriverCreator(output_dir)


@pytest.fixture(scope="module")
def output_dir():
    curr_dir = os.path.dirname(os.path.abspath(__file__))
    output_dir = os.path.abspath(os.path.join(curr_dir, "..", "..", "output_dir"))
    return output_dir


def test_create_chrome_with_options(creator):
    options = mock()
    expected_webdriver = mock()
    when(webdriver).Chrome(
        service_log_path=None, options=options, executable_path="chromedriver"
    ).thenReturn(expected_webdriver)
    driver = creator.create_chrome({}, None, options=options)
    assert driver == expected_webdriver


def test_create_chrome_with_options_and_remote_url(creator):
    url = "http://localhost:4444/wd/hub"
    caps = webdriver.DesiredCapabilities.CHROME.copy()
    options = mock()
    expected_webdriver = mock()
    file_detector = mock_file_detector(creator)
    when(webdriver).Remote(
        command_executor=url,
        desired_capabilities=caps,
        browser_profile=None,
        options=options,
        file_detector=file_detector,
    ).thenReturn(expected_webdriver)
    driver = creator.create_chrome({}, url, options=options)
    assert driver == expected_webdriver


def test_create_headless_chrome_with_options(creator):
    options = mock()
    expected_webdriver = mock()
    when(webdriver).Chrome(
        service_log_path=None, options=options, executable_path="chromedriver"
    ).thenReturn(expected_webdriver)
    driver = creator.create_headless_chrome({}, None, options=options)
    assert driver == expected_webdriver


def test_create_firefox_with_options(creator, output_dir):
    log_file = os.path.join(output_dir, "geckodriver-1.log")
    options = mock()
    profile = mock()
    expected_webdriver = mock()
    when(webdriver).FirefoxProfile().thenReturn(profile)
    when(webdriver).Firefox(
        options=options,
        firefox_profile=profile,
        executable_path="geckodriver",
        service_log_path=log_file,
    ).thenReturn(expected_webdriver)
    driver = creator.create_firefox({}, None, None, options=options)
    assert driver == expected_webdriver


def test_create_firefox_with_options_and_remote_url(creator):
    url = "http://localhost:4444/wd/hub"
    profile = mock()
    when(webdriver).FirefoxProfile().thenReturn(profile)
    caps = webdriver.DesiredCapabilities.FIREFOX.copy()
    options = mock()
    expected_webdriver = mock()
    file_detector = mock_file_detector(creator)
    when(webdriver).Remote(
        command_executor=url,
        desired_capabilities=caps,
        browser_profile=profile,
        options=options,
        file_detector=file_detector,
    ).thenReturn(expected_webdriver)
    driver = creator.create_firefox({}, url, None, options=options)
    assert driver == expected_webdriver


def test_create_headless_firefox_with_options(creator, output_dir):
    log_file = os.path.join(output_dir, "geckodriver-1.log")
    options = mock()
    profile = mock()
    expected_webdriver = mock()
    when(webdriver).FirefoxProfile().thenReturn(profile)
    when(webdriver).Firefox(
        options=options,
        firefox_profile=profile,
        executable_path="geckodriver",
        service_log_path=log_file,
    ).thenReturn(expected_webdriver)
    driver = creator.create_headless_firefox({}, None, None, options=options)
    assert driver == expected_webdriver


def test_create_ie_with_options(creator):
    options = mock()
    expected_webdriver = mock()
    when(webdriver).Ie(
        service_log_path=None, options=options, executable_path="IEDriverServer.exe"
    ).thenReturn(expected_webdriver)
    driver = creator.create_ie({}, None, options=options)
    assert driver == expected_webdriver


def test_create_ie_with_options_and_remote_url(creator):
    url = "http://localhost:4444/wd/hub"
    caps = webdriver.DesiredCapabilities.INTERNETEXPLORER.copy()
    options = mock()
    expected_webdriver = mock()
    file_detector = mock_file_detector(creator)
    when(webdriver).Remote(
        command_executor=url,
        desired_capabilities=caps,
        browser_profile=None,
        options=options,
        file_detector=file_detector,
    ).thenReturn(expected_webdriver)
    driver = creator.create_ie({}, url, options=options)
    assert driver == expected_webdriver


def test_create_ie_with_options_and_log_path(creator):
    options = mock()
    expected_webdriver = mock()
    when(webdriver).Ie(
        options=options, service_log_path=None, executable_path="IEDriverServer.exe"
    ).thenReturn(expected_webdriver)
    driver = creator.create_ie({}, None, options=options)
    assert driver == expected_webdriver


def test_has_options(creator):
    assert creator._has_options(webdriver.Chrome)
    assert creator._has_options(webdriver.Firefox)
    assert creator._has_options(webdriver.Ie)
    assert creator._has_options(webdriver.Edge) is False
    assert creator._has_options(webdriver.Opera)
    assert creator._has_options(webdriver.Safari) is False


def test_create_opera_with_options(creator):
    options = mock()
    expected_webdriver = mock()
    executable_path = "operadriver"
    when(webdriver).Opera(
        options=options, service_log_path=None, executable_path=executable_path
    ).thenReturn(expected_webdriver)
    driver = creator.create_opera({}, None, options=options)
    assert driver == expected_webdriver


def test_create_opera_with_options_and_remote_url(creator):
    url = "http://localhost:4444/wd/hub"
    caps = webdriver.DesiredCapabilities.OPERA.copy()
    options = mock()
    expected_webdriver = mock()
    file_detector = mock_file_detector(creator)
    when(webdriver).Remote(
        command_executor=url,
        desired_capabilities=caps,
        browser_profile=None,
        options=options,
        file_detector=file_detector,
    ).thenReturn(expected_webdriver)
    driver = creator.create_opera({}, url, options=options)
    assert driver == expected_webdriver


def test_create_safari_no_options_support(creator):
    options = mock()
    expected_webdriver = mock()
    executable_path = "/usr/bin/safaridriver"
    when(webdriver).Safari(executable_path=executable_path).thenReturn(
        expected_webdriver
    )
    driver = creator.create_safari({}, None, options=options)
    assert driver == expected_webdriver


def test_create_phantomjs_no_options_support(creator):
    options = mock()
    expected_webdriver = mock()
    executable_path = "phantomjs"
    when(webdriver).PhantomJS(
        service_log_path=None, executable_path=executable_path
    ).thenReturn(expected_webdriver)
    driver = creator.create_phantomjs({}, None, options=options)
    assert driver == expected_webdriver


def test_create_htmlunit_no_options_support(creator):
    caps = webdriver.DesiredCapabilities.HTMLUNIT.copy()
    options = mock()
    expected_webdriver = mock()
    file_detector = mock_file_detector(creator)
    when(webdriver).Remote(
        command_executor="None",
        desired_capabilities=caps,
        browser_profile=None,
        options=options,
        file_detector=file_detector,
    ).thenReturn(expected_webdriver)
    driver = creator.create_htmlunit(
        {"desired_capabilities": caps}, None, options=options
    )
    assert driver == expected_webdriver


def test_create_htmlunit_with_js_no_options_support(creator):
    caps = webdriver.DesiredCapabilities.HTMLUNITWITHJS.copy()
    options = mock()
    expected_webdriver = mock()
    file_detector = mock_file_detector(creator)
    when(webdriver).Remote(
        command_executor="None",
        desired_capabilities=caps,
        browser_profile=None,
        options=options,
        file_detector=file_detector,
    ).thenReturn(expected_webdriver)
    driver = creator.create_htmlunit_with_js({}, None, options=options)
    assert driver == expected_webdriver


def test_android_options_support(creator):
    caps = webdriver.DesiredCapabilities.ANDROID.copy()
    options = mock()
    expected_webdriver = mock()
    file_detector = mock_file_detector(creator)
    when(webdriver).Remote(
        command_executor="None",
        desired_capabilities=caps,
        browser_profile=None,
        options=options,
        file_detector=file_detector,
    ).thenReturn(expected_webdriver)
    driver = creator.create_android({}, None, options=options)
    assert driver == expected_webdriver


def test_iphone_options_support(creator):
    caps = webdriver.DesiredCapabilities.IPHONE.copy()
    options = mock()
    expected_webdriver = mock()
    file_detector = mock_file_detector(creator)
    when(webdriver).Remote(
        command_executor="None",
        desired_capabilities=caps,
        browser_profile=None,
        options=options,
        file_detector=file_detector,
    ).thenReturn(expected_webdriver)
    driver = creator.create_iphone({}, None, options=options)
    assert driver == expected_webdriver


def test_create_driver_chrome(creator):
    str_options = "add_argument:--disable-dev-shm-usage"
    options = mock()
    expected_webdriver = mock()
    when(creator.selenium_options).create("chrome", str_options).thenReturn(options)
    executable_path = "chromedriver"
    when(creator)._get_executable_path(ANY).thenReturn(executable_path)
    when(webdriver).Chrome(
        service_log_path=None, options=options, executable_path=executable_path
    ).thenReturn(expected_webdriver)
    driver = creator.create_driver(
        "Chrome", desired_capabilities={}, remote_url=None, options=str_options
    )
    assert driver == expected_webdriver


def test_create_driver_firefox(creator, output_dir):
    log_file = os.path.join(output_dir, "geckodriver-1.log")
    str_options = "add_argument:--disable-dev-shm-usage"
    options = mock()
    profile = mock()
    when(webdriver).FirefoxProfile().thenReturn(profile)
    expected_webdriver = mock()
    when(creator.selenium_options).create("firefox", str_options).thenReturn(options)
    executable_path = "geckodriver"
    when(creator)._get_executable_path(ANY).thenReturn(executable_path)
    when(webdriver).Firefox(
        options=options,
        firefox_profile=profile,
        executable_path=executable_path,
        service_log_path=log_file,
    ).thenReturn(expected_webdriver)
    driver = creator.create_driver(
        "FireFox", desired_capabilities={}, remote_url=None, options=str_options
    )
    assert driver == expected_webdriver


def mock_file_detector(creator):
    file_detector = mock()
    when(creator)._get_sl_file_detector().thenReturn(file_detector)
    return file_detector
