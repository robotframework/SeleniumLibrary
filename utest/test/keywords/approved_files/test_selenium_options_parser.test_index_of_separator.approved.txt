Get argument index

0) (6, True)
1) (9, False)
2) (6, True)
3) (9, False)
